contrib_bots/lib/giphy/giphy.py

```diff
@@ -0,0 +1,128 @@
+import sys
+import os
+
+def parse_giphy_config(config_file):
+    config = SafeConfigParser()
+    config.readfp(f, config_file)
+    global giphy_api_key
+    giphy_api_key = config.get("giphy-config", "key")
+
+if os.path.exists(os.environ['HOME'] + '/.giphy_config'):
+    with open(os.environ['HOME'] + '/.giphy_config', 'r') as f:
+        parse_giphy_config(f)
+else:
+    print('Giphy bot config file not found, please set up it in ~/.giphy_config\n\n'
+          'Using format:\n\n[giphy-config]\nkey=<giphy API key here>\n\n')
+    sys.exit(1)
+
+
+class GiphyHandler(object):
+    '''
+    This plugin posts a GIF in response to the keywords provided by the user.
+    Images are provided by Giphy, through the public API.
+    The bot looks for messages starting with "@giphy" or @mention of the bot
+    and responds with a message with the GIF based on provided keywords.
+    It also responds to private messages.
+    '''
+    def usage(self):
+        return '''
+            This plugin allows users to post GIFs provided by Giphy.
+            Users should preface keywords with "@giphy" or the Giphy-bot @mention.
+            The bot responds also to private messages.
+            '''
+
+    def triage_message(self, message, client):
+        # To prevent infinite loop in private message, bot will detect
+        # if the sender name is the bot name it will return false.
+        if message['type'] == 'private':
+            return client.full_name != message['sender_full_name']
+
+        # Return True if we want to (possibly) response to this message.
+        original_content = message['content']
+        is_giphy_called = (original_content.startswith('@giphy ') or
+                           message['is_mentioned'])
+
+        return is_giphy_called
+
+    def handle_message(self, message, client, state_handler):
+        bot_response = get_bot_giphy_response(message, client)
+
+        if message['type'] == 'private':
+            client.send_message(dict(
+                type='private',
+                to=message['sender_email'],
+                content=bot_response,
+            ))
+        else:
+            client.send_message(dict(
+                type='stream',
+                to=message['display_recipient'],
+                subject=message['subject'],
+                content=bot_response,
+            ))
+
+
+def get_url_gif_giphy(keyword, api_key):
+    # Return a URL for a Giphy GIF based on keywords given.
+    # In case of error, e.g. failure to fetch a GIF URL, it will
+    # return a number.
+    query = {'s': keyword,
+             'api_key': api_key}
+    try:
+        data = requests.get('http://api.giphy.com/v1/gifs/translate', params=query)
+    except requests.exceptions.ConnectionError as e:  # Usually triggered by bad connection.
+        logging.warning(e)
+        return 1
+
+    search_status = data.json()['meta']['status']
+    if search_status != 200 or data.status_code != 200:
+        return 1
```

Comment:
Instead of returning numbers here and in the next sanity check, you could raise exceptions which would make it more readable in the place where these error may happen.